### ISSUE TITLE
Support for already installed Node

### ISSUE THREAD

**Issue by mischkl** [eirslett/frontend-maven-plugin#471]
It would be nice if there were an option to use the user's own copy of node, if present. We experienced the issue that any native binaries used by NPM packages (e.g. node-sass) are incompatible across node versions. This resulted in the situation that, after executing a frontend build via Maven, all "normal" frontend builds using the system-global Node would crash with an error about incompatible package binaries. Ultimately we had to setup up an optional Maven parameter to specify an isolated Maven frontend build directory, just to avoid this conflict.

In summary: It would be very helpful for the sake of co-existence with machines and coworkers who prefer using their own Node, if it were possible to allow the system Node and NPM versions to take over if present. This could be enabled via an optional setting so that the default behavior of using specific versions would still be enforced.

**Comment by eirslett** [comment]
You could set this up with maven profiles. Have one maven profile that uses frontend-maven-plugin, and one profile that uses exec-maven-plugin.
mvn clean install -PlocalNode would then trigger exec-maven-plugin instead of frontend-maven-plugin.

I'll close this issue as a duplicate of #54 #108 #177 #243 #273 #277 #340 #344 #440 .

**Comment by mischkl** [comment]
Isn't the whole point of this plugin supposed to be that you don't need to use maven-exec plugin, though? Why should I have to create and maintain two completely different maven tasks to accomplish the same thing?

**Comment by eirslett** [comment]
The premise of the plugin is that it's supposed to use an isolated, version-managed installation of node, this plugin promises no more than that. But it comes with a few helper scripts that can actually be used by frontend devs so they run node/npm/webpack/whatever with the version of node/npm that this plugin installs. That should solve your use case. You can find links to these scripts in the README.

The problem itself is not specific to this plugin, or node/npm, it's a general program in software development. If one developer uses Java 7 and another uses Java 8, it will eventually cause problems. That's why you need locked down versions of runtimes, compilers etc.

**Comment by mriehema** [comment]
> The premise of the plugin is that it's supposed to use an **isolated**, version-managed installation of node, this plugin promises no more than that.

This was exactly our problem. every dev used a different version. We needed a simple way to integrate node/npm to our maven repo to work with our CI and with java devs (they don't have any node/npm installed). To avoid any binary problems, our `mvn clean install` deletes all `node `and `node_modules` directories to always get a clean buildable workspace. Our frontend devs uses `npm install`, the java devs `mvn install` and everybody is happy.

**Comment by mischkl** [comment]
I can only respond by illustrating where our problem is coming from. We need to support two use cases, one is the developers and two are the various parties (analysts, product owners, testers...) who "consume" our project on a regular basis directly from the feature branches (we don't have the infrastructure to rely on jenkins/maven to give them pre-packaged wars for every feature branch). Therefore we use this plugin to make things easier for them to build (by virtue of offering a maven frontend build), and easier for us to configure (because the same thing using exec-maven-plugin is a PITA to configure.

On the other hand, we have devs that do primarily frontend work and use a number of custom Grunt tasks that would be overkill, not to mention a serious maintenance overhead, to make them all available from Maven. We don't discourage experimentation with newer versions of Node because it makes it easier to identify and solve incompatibilities ahead of time. Aside from that, even if we were to rely exclusively on the local Node version, we would have to write custom scripts and deal with a special class of potential bugs for every NPM package that expects a CLI to be installed globally (e.g. karma, protractor, webpack, npm itself, and on and on...). All this is much easier to handle if we simply let the frontend devs use the tools the way they were intended, as global packages.

Ultimately it seems like you are telling people with our requirements, which I imagine are the majority of modern frontend devs, to screw off and just don't use your plugin. Doesn't seem very forward thinking.

I guess if you've already ignored 10 issues then I won't convince you either. I suppose this is what forking is for...

**Comment by mriehema** [comment]
Well your usecases sounds like ours:

A) mvn just triggers node/npm to generate resources for CI/devs/retests in a same reproducable stable way. 
B) All other frontend devs use their global installed tools without mvn.

Just make sure mvn clean install will clean `node_modules` to get their node-sass binaries according to the version defined in the `pom.xml`. 

Our config:
```
          <!-- clean node and node_modules in all themes -->
          <plugin>
            <groupId>org.apache.maven.plugins</groupId>
            <artifactId>maven-clean-plugin</artifactId>
            <configuration>
              <filesets>
                <fileset>
                  <directory>${project.basedir}</directory>
                  <includes>
                    <include>**/node_modules/**</include>
                  </includes>
                  <followSymlinks>false</followSymlinks>
                </fileset>
                <fileset>
                  <directory>node</directory>
                </fileset>
              </filesets>
            </configuration>
          </plugin>
          <!-- build themes  -->
        <plugin>
          <groupId>com.github.eirslett</groupId>
          <artifactId>frontend-maven-plugin</artifactId>
          <version>1.0</version>
          <configuration>
            <nodeVersion>v4.5.0</nodeVersion> <!-- LTS -->
            <npmVersion>3.10.6</npmVersion>
          </configuration>
          <executions>...</executions>
        </plugin>
```

If you still need it, feel free to create a pullrequest, which add a parameter/flag to switch to an existing node/npm and skip installing node/npm. But please test it for OSX, Windows and Linux.

**Comment by eirslett** [comment]
That is correct, this plugin is not meant to solve everyone's needs. It only covers a specific use case.

Using the helper scripts, frontend devs would run `./npm run install` instead of `npm install` and `./grunt` instead of `grunt`. Or you could commit a .bashrc that sets it up to use these local scripts directly. Another option; frontend devs could use nvm and use something like .nvmrc to automatically switch to the preferred version of node/npm for that project. If they want to experiment with newer node/npm versions, I suppose that's what feature branches are for?
Maybe the docs should be updated to be clearer that when people use a global installation of node/npm with another version than the maven build requires, they should expect problems.

I'm not saying screw off, I wish you to succeed with your goals, I just don't think this plugin will be able to help you achieve what you're after. If you still feel that forking the project to create a new maven plugin is the best way to move forward, I encourage you to do so.
